htdocs/core/class/html.form.class.php

```diff
@@ -2974,10 +2980,19 @@ public function select_produits_list($selected = 0, $htmlname = 'productid', $fi
 		$reshook = $hookmanager->executeHooks('selectProductsListFrom', $parameters); // Note that $action and $object may have been modified by hook
 		$sql .= $hookmanager->resPrint;
 
-		if (count($warehouseStatusArray)) {
-			$sql .= " LEFT JOIN " . $this->db->prefix() . "product_stock as ps on ps.fk_product = p.rowid";
-			$sql .= " LEFT JOIN " . $this->db->prefix() . "entrepot as e on ps.fk_entrepot = e.rowid AND e.entity IN (" . getEntity('stock') . ")";
-			$sql .= ' AND e.statut IN (' . $this->db->sanitize($this->db->escape(implode(',', $warehouseStatusArray))) . ')'; // Return line if product is inside the selected stock. If not, an empty line will be returned so we will count 0.
+		if (count($warehouseStatusArray) || ((int) $warehouseId) > 0 ) {
```

Comment:
As this is a filter, can you implement it into the WHERE only, using
WHERE EXISTS (....)

Reply:
Not the same result

From sample DB of Dolibarr dolibarr/dev/initdemo/mysqldump_dolibarr_20.0.0.sql

```
SELECT p.rowid,
       p.ref,
       p.label,
       p.description,
       p.barcode,
       p.fk_country,
       p.fk_product_type,
       p.price,
       p.price_ttc,
       p.price_base_type,
       p.tva_tx,
       p.default_vat_code,
       p.duration,
       p.fk_price_expression,
       (CASE WHEN p.stock IS NULL THEN 0 ELSE p.stock END) AS stock,
       pl.label                                            as label_translated,
       pl.description                                      as description_translated
FROM llx_product as p
         LEFT JOIN llx_product_lang as pl ON pl.fk_product = p.rowid AND pl.lang = 'fr_FR'
WHERE p.entity IN (1)
  AND p.fk_product_type = 0
  AND EXISTS (SELECT psw.fk_product FROM llx_product_stock as psw WHERE psw.reel>0 AND psw.fk_entrepot=2) AND ((p.ref LIKE '%COMP%' OR p.label LIKE '%COMP%' OR pl.label LIKE '%COMP%') OR p.barcode LIKE '%COMP%')
ORDER BY p.ref ASC
```

return rows 2 rows
10|COMP-XP4523|Computer XP4523|A powerfull computer XP4523&nbsp;|123456789055|11|0|100|110|HT|10||""||112||
13|COMP-XP4548|Computer XP4523|A powerfull computer XP4523&nbsp;||11|0|100|110|HT|10||""||0|Computer XP4523|A powerfull computer XP4523&nbsp;

But the last row  COMP-XP4548 is not in WAREHOUSEPARIS (id 2)

Can be done with existing SQL or with 
p.rowid IN (....)
if you want in in a where but it's not the better execution paln

Reply:
I disagree. All what you can do with a left join can be done with an EXISTS and/or NOT EXISTS.

Also in your sample you use the filter on p inside the EXISTS instead of into the main WHERE.
The EXISTS or NOT EXISTS must only contains the condition on the table product_stock.

Third point, there is no condition into your sub request EXISTS that is in relation with parent table product so you join is not a join but a cartesian produt.
(SELECT psw.fk_product FROM llx_product_stock as psw WHERE psw.reel>0 AND psw.fk_entrepot=2)
should be 
(SELECT psw.fk_product FROM llx_product_stock as psw WHERE psw.reel>0 AND psw.fk_entrepot=2 AND psw.fk_product = p.rowid)

What about this
SELECT p.rowid,
       p.ref,
       p.label,
       p.description,
       p.barcode,
       p.fk_country,
       p.fk_product_type,
       p.price,
       p.price_ttc,
       p.price_base_type,
       p.tva_tx,
       p.default_vat_code,
       p.duration,
       p.fk_price_expression,
       (CASE WHEN p.stock IS NULL THEN 0 ELSE p.stock END) AS stock,
       pl.label                                            as label_translated,
       pl.description                                      as description_translated
FROM llx_product as p
         LEFT JOIN llx_product_lang as pl ON pl.fk_product = p.rowid AND pl.lang = 'fr_FR'
WHERE p.entity IN (1)
  AND p.fk_product_type = 0 
  AND ((p.ref LIKE '%COMP%' OR p.label LIKE '%COMP%' OR pl.label LIKE '%COMP%') OR p.barcode LIKE '%COMP%')
  AND EXISTS (SELECT psw.fk_product FROM llx_product_stock as psw WHERE psw.reel>0 AND psw.fk_entrepot=2 AND psw.fk_product = p.rowid) 
ORDER BY p.ref ASC

The last AND is a condition to say "product must have stock > 0 into warehouse 2"

Reply:
Of course, I forget the last check on product...

I've update the code, it's more readable like that, just a new `where clause`